### PR TITLE
Local response norm

### PR DESCRIPTION
**Pull request by mansoorcheema**
Implemented LocalResponseNorm and some initial tests for modules and functional. Reference #25883

### CODE REVIEW

File: torch/csrc/api/include/torch/nn/modules/normalization.h [hunk 1]
```diff
@@ -52,5 +52,29 @@ class TORCH_API LayerNormImpl : public torch::nn::Cloneable<LayerNormImpl> {
...
+/// Applies local response normalization over an input signal composed
+/// of several input planes, where channels occupy the second dimension.
+/// Applies normalization across channels
+/// See https://pytorch.org/docs/master/nn.html#torch.nn.LocalResponseNorm to learn
+/// about the exact behavior of this module.
+class TORCH_API LocalResponseNormImpl : public Cloneable<LocalResponseNormImpl> {
+ public:
+  LocalResponseNormImpl(int64_t size)
+      : LocalResponseNormImpl(LocalResponseNormOptions(size)) {}
+  explicit LocalResponseNormImpl(const LocalResponseNormOptions & options_ );
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/modules/normalization.h]: nit:
```suggestion
  explicit LocalResponseNormImpl(const LocalResponseNormOptions& options_);
```

File: torch/csrc/api/include/torch/nn/modules/normalization.h [hunk 2]
```diff
@@ -1,8 +1,8 @@
 #pragma once
 
-#include <torch/nn/cloneable.h>
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/modules/normalization.h]: Thanks for catching this! :D

File: torch/csrc/api/include/torch/nn/modules/normalization.h [hunk 3]
```diff
@@ -1,8 +1,8 @@
...
 #include <torch/nn/cloneable.h>
 #include <torch/nn/functional/normalization.h>
+#include <torch/nn/functional/pooling.h>
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/modules/normalization.h]: Is there a specific reason we need `#include <torch/nn/functional/pooling.h>` here?

File: torch/csrc/api/include/torch/nn/functional/normalization.h [hunk 4]
```diff
@@ -10,7 +15,7 @@ inline Tensor normalize(
     const Tensor& input,
     const NormalizeOptions& options = {},
     c10::optional<Tensor> out = c10::nullopt) {
-
+
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/functional/normalization.h]: nit: we can likely remove this line

File: torch/csrc/api/include/torch/nn/functional/normalization.h [hunk 5]
```diff
@@ -28,6 +33,26 @@ inline Tensor layer_norm(const Tensor& input,
     return torch::layer_norm(input, options.normalized_shape(), weight, bias, options.eps());
 }
 
+inline Tensor local_response_norm(
+    const Tensor& input,
+    const LocalResponseNormOptions& options) {
+    auto dim = input.dim();
+    TORCH_CHECK( dim >=3 ,"Expected 3D or higher dimensionality input (got ",dim, " ,dimensions)");
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/functional/normalization.h]: nit:
```suggestion
    TORCH_CHECK(dim >= 3, "Expected 3D or higher dimensionality input (got ", dim, " dimensions)");
```

File: torch/csrc/api/include/torch/nn/functional/normalization.h [hunk 6]
```diff
@@ -28,6 +33,26 @@ inline Tensor layer_norm(const Tensor& input,
...
+    auto div = input.mul(input).unsqueeze(1);
+    if (dim == 3) {
+      div = pad(div, PadOptions({0, 0, options.size() / 2, (options.size() - 1) /2}));
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/functional/normalization.h]: nit:
```suggestion
      div = pad(div, PadOptions({0, 0, options.size() / 2, (options.size() - 1) / 2}));
```

File: torch/csrc/api/include/torch/nn/functional/normalization.h [hunk 7]
```diff
@@ -1,6 +1,11 @@
 #pragma once
 
 #include <torch/nn/options/normalization.h>
+#include <torch/nn/functional/padding.h>
+#include <torch/nn/functional/pooling.h>
+#include <torch/types.h>
+
+#include <vector>
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/functional/normalization.h]: we can likely remove this include

File: test/cpp/api/modules.cpp [hunk 8]
```diff
@@ -768,6 +768,37 @@ TEST_F(ModulesTest, Linear) {
   }
 }
 
+TEST_F(ModulesTest, LocalResponseNorm) {
+  {
+    LocalResponseNorm model(LocalResponseNormOptions(2));
+    auto x = torch::randn({32, 5, 24, 24}, torch::requires_grad());
+    auto y = model(x);
+    //auto y_exp = torch::layer_norm(x, {2, 2}, model->weight, model->bias, 2e-5);
```
Comment (yf225) [inline test/cpp/api/modules.cpp]: For the expected outputs in the module tests, I think we can get the numerical values from Python and directly compare to those hardcoded values :D

**Review by yf225 — COMMENTED** [review]
@mansoorcheema Thanks so much for the awesome work! I left some minor comments.

File: torch/csrc/api/include/torch/nn/modules/normalization.h (hunk 3 above)
Comment (mansoorcheema) [reply torch/csrc/api/include/torch/nn/modules/normalization.h]: Yes, its note needed here. I shall remove this.

**Review by mansoorcheema — COMMENTED** [review]
Updated with the suggested changes and module test.

**Comment by mansoorcheema** [comment]
@yf225 I updated with the code with suggested changes and module test with expected values from python API. Let me know what you think.

**Review by yf225 — APPROVED** [review]
@mansoorcheema Thanks so much for the awesome work!
